### PR TITLE
Treat _retArg as a single def in inferConst

### PR DESCRIPTION
**Pull request by benharsh**
The compiler currently transforms functions that return certain types
(e.g., tuples, records) such that they will return through a reference
formal instead of a PRIM_RETURN. Such formals should not be considered
to have the const-ref intent, as they will modify the thing they are
pointing to.

The inferConst function, however, can consider this to be a single 'def'
of a value. This may allow for the value to become a const-val, which
can allow RVF to be more effective.

Testing:
- [x] local
- [x] no-local
- [x] gasnet, dist-cyclic, dist-replicated

### CODE REVIEW

**Comment by benharsh** [comment]
For those interested, I expect this to resolve the performance regression in the HPL release version for 16-node-xc.

File: compiler/optimizations/inferConstRefs.cpp [hunk 1]
```diff
@@ -499,8 +499,33 @@ static bool inferConst(Symbol* sym) {
...
+            if (lhsFormal->hasFlag(FLAG_RETARG) == false) {
+              isRetArg = false;
+            }
+          } else {
+            isRetArg = false;
+          }
+        }
+
+        if (isRetArg) {
+          numDefs += 1;
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Also, I'm a little bit confused about the interplay between the ArgSymbol case above and the case here dealing with the result of ADDR_OF / SET_REFERENCE.  Could this have been computed in the ArgSymbol part? Would that make it simpler?

Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Should the if (numDefs > 1) isConstVal = false move to the bottom of the function so as to be shared by the different cases?

Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: Yikes, I really overlooked the numDefs issue... I'll work on moving the check to the bottom of the function.

I think we do need to handle the ADDR_OF/SET_REFERENCE case to catch this pattern:
```
ref ret_to_arg_ref_tmp = (addrof sym);
call buildSomeType(..., ret_to_arg_ref_tmp);
```

Though having said that I could also add a check in the ArgSymbol case if we pass the symbol without an explicit addrf.

File: compiler/optimizations/inferConstRefs.cpp [hunk 2]
```diff
@@ -499,8 +499,33 @@ static bool inferConst(Symbol* sym) {
...
-        SymExpr* LHS = toSymExpr(parent->get(1));
-        if (LHS->isRef() && !inferConstRef(LHS->symbol())) {
+        Symbol* LHS = toSymExpr(parent->get(1))->symbol();
+        INT_ASSERT(LHS->isRef());
+
+        //
+        // If 'sym' is constructed through a _retArg, we can consider that to
+        // be a single 'def'.
+        //
+        bool isRetArg = true;
```
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: Yes, that's a good idea.

**Comment by benharsh** [comment]
Testing for the most recent commit has come back clean. How do these changes look?

**Comment by mppf** [comment]
looks much better, thanks.

**Pull request merged**